README.md
# py-openweather-tray

System tray app for written in Python that shows outside temperature in given city 
using data from openweathermap.org. 
Color of text is changing based on the temperature.
Supports both Celsius and Fahrenheit scale.

Tooltip contains additional weather info (wind speed, direction, sunrise, sunset, etc.)

Screenshot:
<img src="images/screenshot.png">

## Motivation

I wanted to compare very similar application that I wrote in Clojure (clj-openweather-tray) with this one.

## Findings
I must say that even though I like to write code in Clojure, it took me longer to write clj-openweather-tray app
and Python code is much more readable.

16 pixels, which is the size of tray icon on Windows OS, is not much if you want to fit 3 or 4 chars inside and still
be able to read it :)

## Build

This project is using [fbs](https://github.com/mherrmann/fbs-tutorial).

Create a virtual environment in the current directory:

    python3 -m venv venv

Activate the virtual environment:

    # On Mac/Linux:
    source venv/bin/activate
    # On Windows:
    call venv\scripts\activate.bat
    # On Windows in PowerShell
    .\venv\scripts\activate.ps1

The remainder of the tutorial assumes that the virtual environment is active.

Install the required libraries:

    pip install -r requirements.txt

(If this produces errors, try `pip install wheel` first.)

## Run the app

    fbs run
   
## Installation

See [fbs tutorial](https://github.com/mherrmann/fbs-tutorial).

## License

Copyright © 2020 Miroslav Havrlent

This program and the accompanying materials are made available under the
terms of the Eclipse Public License 2.0 which is available at
http://www.eclipse.org/legal/epl-2.0.

This Source Code may also be made available under the following Secondary
Licenses when the conditions for such availability set forth in the Eclipse
Public License, v. 2.0 are satisfied: GNU General Public License as published by
the Free Software Foundation, either version 2 of the License, or (at your
option) any later version, with the GNU Classpath Exception which is available
at https://www.gnu.org/software/classpath/license.html.


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file. This change log follows the conventions of [keepachangelog.com](http://keepachangelog.com/).

## [0.0.1] - 2020-07-24
### Added
- first version

tests/test_main.py
from src.main.python import main

def test_get_weather_data():
    assert (main.get_weather_data(
        'https://samples.openweathermap.org/data/2.5/weather?id=2172797&appid=b6907d289e10d714a6e88b30761fae22')
            ['main']['temp'] == 300.15)


def test_kelvin_to_celsius():
    assert (main.kelvin_to_celsius(1) == -272.15)
    assert (main.kelvin_to_celsius(273.15) == 0.0)


def test_kelvin_to_fahrenheit():
    assert (main.kelvin_to_fahrenheit(1) == -457.8714388489209)
    assert (main.kelvin_to_fahrenheit(273.15) == 31.606978417266077)


def test_convert_temperature_test():
    assert (main.convert_temperature(1, 'C') == -272.15)
    assert (main.convert_temperature(1, 'F') == -457.8714388489209)


def test_deg_to_cardinal():
    assert (main.deg_to_cardinal(0) == 'N')
    assert (main.deg_to_cardinal(25) == 'NNE')
    assert (main.deg_to_cardinal(50) == 'NE')
    assert (main.deg_to_cardinal(75) == 'ENE')
    assert (main.deg_to_cardinal(100) == 'E')
    assert (main.deg_to_cardinal(110) == 'ESE')
    assert (main.deg_to_cardinal(125) == 'SE')
    assert (main.deg_to_cardinal(150) == 'SSE')
    assert (main.deg_to_cardinal(175) == 'S')
    assert (main.deg_to_cardinal(200) == 'SSW')
    assert (main.deg_to_cardinal(225) == 'SW')
    assert (main.deg_to_cardinal(250) == 'WSW')
    assert (main.deg_to_cardinal(275) == 'W')
    assert (main.deg_to_cardinal(300) == 'WNW')
    assert (main.deg_to_cardinal(325) == 'NW')
    assert (main.deg_to_cardinal(335) == 'NNW')
    assert (main.deg_to_cardinal(350) == 'N')
    assert (main.deg_to_cardinal(360) == 'N')
    assert (main.deg_to_cardinal(365) == 'N')
    assert (main.deg_to_cardinal(385) == 'NNE')


def test_clouds_percentage_to_text():
    assert (main.clouds_percentage_to_text(5) == 'Clear sky')
    assert (main.clouds_percentage_to_text(86) == 'Overcast clouds')


def test_extract_hour_and_minute_from_epoch():
    assert (main.extract_hour_and_minute_from_epoch(1585283921) == '04:38 UTC')


def test_get_color():
    colors = {'0': '#9bbcff', '223': '#9bbcff', '228': '#9dbdff',
              '233': '#9dbeff', '238': '#9fbfff', '243': '#a1c0ff',
              '248': '#a5c2ff', '253': '#a8c5ff', '258': '#afc9ff',
              '263': '#b9d0ff', '268': '#cedcff', '273': '#fef9ff',
              '278': '#fff3ef', '283': '#ffece0', '288': '#ffe3ca',
              '293': '#ffd9b6', '298': '#ffcc99', '303': '#ffbe7e',
              '308': '#ffad5e', '313': '#ff932c', '318': '#ff7300',
              '323': '#ff3300'}
    assert (main.get_color(colors, -100) == '#9bbcff')
    assert (main.get_color(colors, 0) == '#9bbcff')
    assert (main.get_color(colors, 100) == '#9bbcff')
    assert (main.get_color(colors, 115) == '#9bbcff')
    assert (main.get_color(colors, 226) == '#9dbdff')
    assert (main.get_color(colors, 238) == '#9fbfff')
    assert (main.get_color(colors, 9999) == '#ff3300')


def test_find_closest():
    assert (main.find_closest([1, 2, 3, 10, 20, 50, 100], 5) == 3)
    assert (main.find_closest([1, 2, 3, 10, 20, 50, 100], -100) == 1)
    assert (main.find_closest([1, 2, 3, 10, 20, 50, 100], 9999) == 100)


src/main/python/main.py
from fbs_runtime.application_context.PyQt5 import ApplicationContext
from PyQt5 import QtWidgets, QtCore, QtGui
import sys
import requests
import json
from datetime import datetime
from bisect import bisect_left


class SystemTrayIcon(QtWidgets.QSystemTrayIcon):

    def __init__(self, config, parent=None):
        QtWidgets.QSystemTrayIcon.__init__(self, parent)
        menu = QtWidgets.QMenu(parent)
        menu.addAction('Exit')
        self.setContextMenu(menu)
        menu.triggered.connect(self.exit)
        self.config = config
        self._update_secs = config.get('update_secs')
        self._start_update_timer()

    def do_update(self):
        update_tray(self.config, self)
        self._start_update_timer()

    def _start_update_timer(self):
        try:
            self._updateTimer.stop()
        except:
            pass
        self._updateTimer = QtCore.QTimer()
        self._updateTimer.timeout.connect(self.do_update)
        self._updateTimer.start(self._update_secs * 1000)

    def exit(self):
        QtCore.QCoreApplication.exit()


def deg_to_cardinal(deg):
    if deg is not None:
        points = ['N', 'NNE', 'NE', 'ENE', 'E', 'ESE', 'SE', 'SSE',
                  'S', 'SSW', 'SW', 'WSW', 'W', 'WNW', 'NW', 'NNW']
        index = round(deg / (360. / len(points)))
        return points[index % len(points)]
    else:
        return 'Data missing'


def clouds_percentage_to_text(perc):
    if (perc >= 0) and (perc <= 10):
        return 'Clear sky'
    if (perc >= 11) and (perc <= 24):
        return 'Few clouds'
    if (perc >= 25) and (perc <= 50):
        return 'Scattered clouds'
    if (perc >= 51) and (perc <= 84):
        return 'Broken clouds'
    if (perc >= 85) and (perc <= 100):
        return 'Overcast clouds'


def extract_hour_and_minute_from_epoch(epoch):
    return datetime.utcfromtimestamp(epoch).strftime('%H:%M UTC')


def create_tooltip_text(wind_speed, wind_deg, humidity, clouds, sunrise, sunset):
    str_list = []
    str_list.append('Wind speed: %s m/s\n' % wind_speed)
    str_list.append('Wind direction: %s (%s)\n' %
                    (wind_deg, deg_to_cardinal(wind_deg)))
    str_list.append('Humidity: %s %%\n' % humidity)
    str_list.append('Sky: %s\n' % clouds_percentage_to_text(clouds))
    str_list.append('Sunrise: %s\n' %
                    extract_hour_and_minute_from_epoch(sunrise))
    str_list.append('Sunset: %s' % extract_hour_and_minute_from_epoch(sunset))
    return ''.join(str_list)


def get_weather_data(url):
    response = requests.get(url)
    return response.json()


def kelvin_to_celsius(kelvin):
    return kelvin - 273.15


def kelvin_to_fahrenheit(kelvin):
    return (kelvin / 0.556) - 459.67


def convert_temperature(temp, scale):
    if scale == 'C':
        return kelvin_to_celsius(temp)
    if scale == 'F':
        return kelvin_to_fahrenheit(temp)


def get_color(colors, temp):
    colors_keys = list(map(int, [*colors.keys()]))
    colors_keys.sort()
    closest_key = find_closest(colors_keys, temp)
    closest_value = colors.get('%s' % closest_key)
    return closest_value


def find_closest(sl, v):
    """
    Assumes sl is a sorted list. Returns closest value to v.
    If two numbers are equally close, return the smallest number.
    """
    pos = bisect_left(sl, v)
    if pos == 0:
        return sl[0]
    if pos == len(sl):
        return sl[-1]
    before = sl[pos - 1]
    after = sl[pos]
    if after - v < v - before:
        return after
    else:
        return before


def draw_tray_icon(tray, text, color, font, font_size, icon_size, tooltip_text):
    px = QtGui.QPixmap(icon_size, icon_size)
    px.fill(QtCore.Qt.transparent)
    painter = QtGui.QPainter(px)
    pen_color = QtGui.QColor(color)
    pen = QtGui.QPen(pen_color)
    painter.setPen(pen)
    painter.setFont(QtGui.QFont(font, font_size))
    painter.drawText(px.rect(), QtCore.Qt.AlignCenter, text)
    painter.end()
    qicon = QtGui.QIcon()
    qicon.addPixmap(px)
    tray.setIcon(qicon)
    tray.setToolTip(tooltip_text)


def update_tray(config, tray):
    url = config.get(
        'url', 'https://samples.openweathermap.org/data/2.5/weather?id=2172797&appid=b6907d289e10d714a6e88b30761fae22')
    weather_data = get_weather_data(url)
    temp = weather_data['main']['temp']
    converted_temp = round(convert_temperature(temp, config.get('scale', 'C')))
    draw_tray_icon(tray,
                   str(converted_temp),
                   get_color(config.get('colors'), temp),
                   config.get('font', 'Arial'),
                   int(config.get('font_size', 8)),
                   int(config.get('icon_size', 16)),
                   create_tooltip_text(weather_data['wind']['speed'],
                                       weather_data['wind']['deg'],
                                       weather_data['main']['humidity'],
                                       weather_data['clouds']['all'],
                                       weather_data['sys']['sunrise'],
                                       weather_data['sys']['sunset']))


def main():
    appctxt = ApplicationContext()
    w = QtWidgets.QWidget()
    with open('config.json') as json_file:
        config = json.load(json_file)
    tray = SystemTrayIcon(config, w)
    update_tray(config, tray)
    tray.show()
    exit_code = appctxt.app.exec_()
    sys.exit(exit_code)


if __name__ == '__main__':
    main()
